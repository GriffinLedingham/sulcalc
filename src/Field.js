import { Terrains, Weathers, maxGen } from "./utilities";

export default class Field {
  constructor(field = {}) {
    this.gen = Number(field.gen ?? maxGen);
    this.multiBattle = Boolean(field.multiBattle);
    this.invertedBattle = Boolean(field.invertedBattle);

    this.weather = Number(field.weather ?? Weathers.CLEAR);
    this.airLock = Boolean(field.airLock);

    this.waterSport = Boolean(field.waterSport);
    this.mudSport = Boolean(field.mudSport);

    this.magicRoom = Boolean(field.magicRoom);
    this.wonderRoom = Boolean(field.wonderRoom);
    this.gravity = Boolean(field.gravity);

    this.terrain = field.terrain ?? Terrains.NO_TERRAIN;

    this.fairyAura = Boolean(field.fairyAura);
    this.darkAura = Boolean(field.darkAura);
    this.auraBreak = Boolean(field.auraBreak);

    this.ionDeluge = Boolean(field.ionDeluge);
  }

  effectiveWeather() {
    return this.airLock ? Weathers.CLEAR : this.weather;
  }

  isClearWeather() {
    return this.effectiveWeather() === Weathers.CLEAR;
  }

  rain() {
    return [Weathers.RAIN, Weathers.HEAVY_RAIN].includes(
      this.effectiveWeather()
    );
  }

  sun() {
    return [Weathers.SUN, Weathers.HARSH_SUN].includes(this.effectiveWeather());
  }

  sand() {
    return this.effectiveWeather() === Weathers.SAND;
  }

  hail() {
    return this.effectiveWeather() === Weathers.HAIL;
  }

  heavyRain() {
    return this.effectiveWeather() === Weathers.HEAVY_RAIN;
  }

  harshSun() {
    return this.effectiveWeather() === Weathers.HARSH_SUN;
  }

  strongWinds() {
    return this.effectiveWeather() === Weathers.STRONG_WINDS;
  }

  grassyTerrain() {
    return this.terrain === Terrains.GRASSY_TERRAIN;
  }

  mistyTerrain() {
    return this.terrain === Terrains.MISTY_TERRAIN;
  }

  electricTerrain() {
    return this.terrain === Terrains.ELECTRIC_TERRAIN;
  }

  psychicTerrain() {
    return this.terrain === Terrains.PSYCHIC_TERRAIN;
  }
}
