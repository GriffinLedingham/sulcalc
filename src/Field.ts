import { Generation, Terrain, Weather, maxGen } from "./utilities";

export type FieldOptions = Partial<Field>;

export default class Field {
  public gen: Generation = maxGen;
  public weather: Weather = Weather.CLEAR;
  public terrain: Terrain = Terrain.NO_TERRAIN;
  public multiBattle: boolean = false;
  public invertedBattle: boolean = false;
  public airLock: boolean = false;
  public waterSport: boolean = false;
  public mudSport: boolean = false;
  public magicRoom: boolean = false;
  public wonderRoom: boolean = false;
  public gravity: boolean = false;
  public fairyAura: boolean = false;
  public darkAura: boolean = false;
  public auraBreak: boolean = false;
  public ionDeluge: boolean = false;

  public constructor(field: Field | FieldOptions = {}) {
    Object.assign(this, field);
  }

  public effectiveWeather(): Weather {
    return this.airLock ? Weather.CLEAR : this.weather;
  }

  public isClearWeather(): boolean {
    return this.effectiveWeather() === Weather.CLEAR;
  }

  public rain(): boolean {
    const weather = this.effectiveWeather();
    return weather === Weather.RAIN || weather === Weather.HEAVY_RAIN;
  }

  public sun(): boolean {
    const weather = this.effectiveWeather();
    return weather === Weather.SUN || weather === Weather.HARSH_SUN;
  }

  public sand(): boolean {
    return this.effectiveWeather() === Weather.SAND;
  }

  public hail(): boolean {
    return this.effectiveWeather() === Weather.HAIL;
  }

  public heavyRain(): boolean {
    return this.effectiveWeather() === Weather.HEAVY_RAIN;
  }

  public harshSun(): boolean {
    return this.effectiveWeather() === Weather.HARSH_SUN;
  }

  public strongWinds(): boolean {
    return this.effectiveWeather() === Weather.STRONG_WINDS;
  }

  public grassyTerrain(): boolean {
    return this.terrain === Terrain.GRASSY_TERRAIN;
  }

  public mistyTerrain(): boolean {
    return this.terrain === Terrain.MISTY_TERRAIN;
  }

  public electricTerrain(): boolean {
    return this.terrain === Terrain.ELECTRIC_TERRAIN;
  }

  public psychicTerrain(): boolean {
    return this.terrain === Terrain.PSYCHIC_TERRAIN;
  }
}
