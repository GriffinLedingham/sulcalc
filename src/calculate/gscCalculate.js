import { isPhysicalType, isSpecialType } from "../info";
import {
  Stats,
  Types,
  damageVariation,
  scaleStat,
  needsScaling
} from "../utilities";
import moveInfo from "./moveInfo";

const typeToPresent = {
  [Types.NORMAL]: 0,
  [Types.FIGHTING]: 1,
  [Types.FLYING]: 2,
  [Types.POISON]: 3,
  [Types.GROUND]: 4,
  [Types.ROCK]: 5,
  [Types.BUG]: 7,
  [Types.GHOST]: 8,
  [Types.STEEL]: 9,
  [Types.FIRE]: 20,
  [Types.WATER]: 21,
  [Types.GRASS]: 22,
  [Types.ELECTRIC]: 23,
  [Types.PSYCHIC]: 24,
  [Types.ICE]: 25,
  [Types.DRAGON]: 26,
  [Types.DARK]: 27
};

export default (attacker, defender, move, field) => {
  const { moveType, movePower, effectiveness, fail } = moveInfo(
    attacker,
    defender,
    move,
    field
  );
  if (fail || effectiveness[0] === 0) return [0];

  let level = attacker.level;
  const atkBoost = attacker.boost(Stats.ATK);
  const defBoost = defender.boost(Stats.DEF);
  const satkBoost = attacker.boost(Stats.SATK);
  const sdefBoost = defender.boost(Stats.SDEF);

  let atk, def;
  if (move.critical && atkBoost <= defBoost) {
    atk = attacker.stat(Stats.ATK);
    def = defender.stat(Stats.DEF);
  } else {
    atk = attacker.boostedStat(Stats.ATK);
    def = defender.boostedStat(Stats.DEF);
    if (attacker.isBurned()) atk = Math.trunc(atk / 2);
    if (defender.reflect) def *= 2;
  }

  let satk, sdef;
  if (move.critical && satkBoost <= sdefBoost) {
    satk = attacker.stat(Stats.SATK);
    sdef = defender.stat(Stats.SDEF);
  } else {
    satk = attacker.boostedStat(Stats.SATK);
    sdef = defender.boostedStat(Stats.SDEF);
    if (defender.lightScreen) sdef *= 2;
  }

  if (attacker.thickClubBoosted()) atk *= 2;
  if (attacker.lightBallBoosted()) satk *= 2;

  let a, d;
  if (isPhysicalType(moveType)) {
    a = atk;
    d = def;
  } else if (isSpecialType(moveType)) {
    a = satk;
    d = sdef;
  } else {
    return [0];
  }

  if (needsScaling(a, d)) {
    a = scaleStat(a);
    d = Math.max(1, scaleStat(d));
  }
  // in-game Crystal would repeatedly shift by 2 until it fits in a byte

  if (attacker.name === "Ditto" && attacker.item.name === "Metal Powder") {
    d = Math.trunc(d * 3 / 2);
    if (needsScaling(d)) {
      a = scaleStat(a, 1);
      d = Math.max(1, scaleStat(d, 1));
    }
  }

  if (move.isExplosion()) {
    d = Math.max(1, Math.trunc(d / 2));
  }

  if (move.name === "Beat Up") {
    a = attacker.beatUpStats[move.beatUpHit];
    d = defender.baseStat(Stats.DEF);
    level = attacker.beatUpLevels[move.beatUpHit];
  } else if (move.name === "Present") {
    a = Math.floor(10 * effectiveness[0] / effectiveness[1]);
    // Intentionally switches the attacker and defender's secondary types
    d = typeToPresent[attacker.secondaryType()] || 0;
    level = typeToPresent[defender.secondaryType()] || 0;
  }

  d = Math.max(1, d);
  let baseDamage = Math.trunc(
    Math.trunc(Math.trunc(2 * level / 5 + 2) * movePower * a / d) / 50
  );

  if (move.critical) {
    baseDamage *= 2;
  }

  if (attacker.item.boostedType() === moveType) {
    baseDamage = Math.trunc(baseDamage * 110 / 100);
  }

  baseDamage = Math.min(997, baseDamage) + 2;

  if (field.sun()) {
    if (moveType === Types.FIRE) {
      baseDamage = Math.trunc(baseDamage * 3 / 2);
    } else if (moveType === Types.WATER) {
      baseDamage = Math.trunc(baseDamage / 2);
    }
  } else if (field.rain()) {
    if (moveType === Types.WATER) {
      baseDamage = Math.trunc(baseDamage * 3 / 2);
    } else if (moveType === Types.FIRE || move.name === "Solar Beam") {
      baseDamage = Math.trunc(baseDamage / 2);
    }
  }

  if (attacker.stab(moveType)) {
    baseDamage = Math.trunc(baseDamage * 3 / 2);
  }

  baseDamage = Math.trunc(baseDamage * effectiveness[0] / effectiveness[1]);

  // these don't have damage variance
  if (move.name === "Reversal" || move.name === "Flail") {
    return [baseDamage];
  }

  let damages = damageVariation(baseDamage, 217, 255);

  if (move.name === "Pursuit" && defender.switchedOut) {
    damages = damages.map(d => 2 * d);
  }

  return damages;
};
