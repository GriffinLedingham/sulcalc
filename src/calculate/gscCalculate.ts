import Field from "../Field";
import { isPhysicalType, isSpecialType } from "../info";
import Move from "../Move";
import Pokemon from "../Pokemon";
import {
  Stat,
  Type,
  damageVariation,
  needsScaling,
  scaleStat
} from "../utilities";
import moveInfo from "./moveInfo";

const typeToPresent: { [key in Type]?: number } = {
  [Type.NORMAL]: 0,
  [Type.FIGHTING]: 1,
  [Type.FLYING]: 2,
  [Type.POISON]: 3,
  [Type.GROUND]: 4,
  [Type.ROCK]: 5,
  [Type.BUG]: 7,
  [Type.GHOST]: 8,
  [Type.STEEL]: 9,
  [Type.FIRE]: 20,
  [Type.WATER]: 21,
  [Type.GRASS]: 22,
  [Type.ELECTRIC]: 23,
  [Type.PSYCHIC]: 24,
  [Type.ICE]: 25,
  [Type.DRAGON]: 26,
  [Type.DARK]: 27
};

export default (
  attacker: Pokemon,
  defender: Pokemon,
  move: Move,
  field: Field
): number[] => {
  const info = moveInfo(attacker, defender, move, field);
  if (info.fail || info.effectiveness[0] === 0) return [0];
  const { moveType, movePower, effectiveness } = info;

  let level = attacker.level;
  const atkBoost = attacker.boost(Stat.ATK);
  const defBoost = defender.boost(Stat.DEF);
  const satkBoost = attacker.boost(Stat.SATK);
  const sdefBoost = defender.boost(Stat.SDEF);

  let atk, def;
  if (move.critical && atkBoost <= defBoost) {
    atk = attacker.stat(Stat.ATK);
    def = defender.stat(Stat.DEF);
  } else {
    atk = attacker.boostedStat(Stat.ATK);
    def = defender.boostedStat(Stat.DEF);
    if (attacker.isBurned()) atk = Math.trunc(atk / 2);
    if (defender.reflect) def *= 2;
  }

  let satk, sdef;
  if (move.critical && satkBoost <= sdefBoost) {
    satk = attacker.stat(Stat.SATK);
    sdef = defender.stat(Stat.SDEF);
  } else {
    satk = attacker.boostedStat(Stat.SATK);
    sdef = defender.boostedStat(Stat.SDEF);
    if (defender.lightScreen) sdef *= 2;
  }

  if (attacker.thickClubBoosted()) atk *= 2;
  if (attacker.lightBallBoosted()) satk *= 2;

  let a, d;
  if (isPhysicalType(moveType)) {
    a = atk;
    d = def;
  } else if (isSpecialType(moveType)) {
    a = satk;
    d = sdef;
  } else {
    return [0];
  }

  if (needsScaling(a, d)) {
    a = scaleStat(a);
    d = Math.max(1, scaleStat(d));
  }
  // in-game Crystal would repeatedly shift by 2 until it fits in a byte

  if (attacker.name === "Ditto" && attacker.item.name === "Metal Powder") {
    d = Math.trunc((d * 3) / 2);
    if (needsScaling(d)) {
      a = scaleStat(a, 1);
      d = Math.max(1, scaleStat(d, 1));
    }
  }

  if (move.isExplosion()) {
    d = Math.max(1, Math.trunc(d / 2));
  }

  if (move.name === "Beat Up") {
    a = attacker.beatUpStats[move.beatUpHit];
    d = defender.baseStat(Stat.DEF);
    level = attacker.beatUpLevels[move.beatUpHit];
  } else if (move.name === "Present") {
    a = Math.floor((10 * effectiveness[0]) / effectiveness[1]);
    // intentionally switches the attacker and defender's secondary types
    d = typeToPresent[attacker.secondaryType()] || 0;
    level = typeToPresent[defender.secondaryType()] || 0;
  }

  d = Math.max(1, d);
  let baseDamage = Math.trunc(
    Math.trunc((Math.trunc((2 * level) / 5 + 2) * movePower * a) / d) / 50
  );

  if (move.critical) {
    baseDamage *= 2;
  }

  if (attacker.item.boostedType() === moveType) {
    baseDamage = Math.trunc((baseDamage * 110) / 100);
  }

  baseDamage = Math.min(997, baseDamage) + 2;

  if (field.sun()) {
    if (moveType === Type.FIRE) {
      baseDamage = Math.trunc((baseDamage * 3) / 2);
    } else if (moveType === Type.WATER) {
      baseDamage = Math.trunc(baseDamage / 2);
    }
  } else if (field.rain()) {
    if (moveType === Type.WATER) {
      baseDamage = Math.trunc((baseDamage * 3) / 2);
    } else if (moveType === Type.FIRE || move.name === "Solar Beam") {
      baseDamage = Math.trunc(baseDamage / 2);
    }
  }

  if (attacker.stab(moveType)) {
    baseDamage = Math.trunc((baseDamage * 3) / 2);
  }

  baseDamage = Math.trunc((baseDamage * effectiveness[0]) / effectiveness[1]);

  // these don't have damage variance
  if (move.name === "Reversal" || move.name === "Flail") {
    return [baseDamage];
  }

  let damages = damageVariation(baseDamage, 217, 255);

  if (move.name === "Pursuit" && defender.switchedOut) {
    damages = damages.map(d => 2 * d);
  }

  return damages;
};
