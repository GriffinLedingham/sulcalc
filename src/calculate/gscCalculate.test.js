import gscCalculate from "sulcalc/calculate/gscCalculate";
import Pokemon from "sulcalc/Pokemon";
import Move from "sulcalc/Move";
import Field from "sulcalc/Field";
import { Gens, Stats, Statuses, Weathers } from "sulcalc/utilities";

const gen = Gens.GSC;

let field, rain, sand, sun;
beforeEach(() => {
  field = new Field({ gen });
  rain = new Field({ weather: Weathers.RAIN, gen });
  sand = new Field({ weather: Weathers.SAND, gen });
  sun = new Field({ weather: Weathers.SUN, gen });
});

describe("crit mechanics", () => {
  let machamp, starmie, crossChopCrit;
  beforeEach(() => {
    machamp = new Pokemon({ name: "Machamp", gen });
    starmie = new Pokemon({ name: "Starmie", gen });
    crossChopCrit = new Move({
      name: "Cross Chop",
      critical: true,
      gen
    });
  });

  test("doubles damage", () => {
    const damage = gscCalculate(machamp, starmie, crossChopCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(168);
  });

  test("only ignores modifiers if boosts are the same or worse", () => {
    machamp.boosts[Stats.ATK] = 6;
    starmie.boosts[Stats.DEF] = 5;
    let damage = gscCalculate(machamp, starmie, crossChopCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(135);

    machamp.boosts[Stats.ATK] = 6;
    starmie.boosts[Stats.DEF] = 6;
    damage = gscCalculate(machamp, starmie, crossChopCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(168);

    machamp.boosts[Stats.ATK] = 5;
    starmie.boosts[Stats.DEF] = 6;
    damage = gscCalculate(machamp, starmie, crossChopCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(168);
  });

  test("ignoring includes Reflect", () => {
    starmie.reflect = true;

    let damage = gscCalculate(machamp, starmie, crossChopCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(168);

    machamp.boosts[Stats.ATK] = 1;
    damage = gscCalculate(machamp, starmie, crossChopCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(127);
  });

  test("ignoring includes Light Screen", () => {
    starmie.lightScreen = true;
    const thunderCrit = new Move({
      name: "Thunder",
      critical: true,
      gen
    });

    let damage = gscCalculate(machamp, starmie, thunderCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(344);

    machamp.boosts[Stats.SATK] = 1;
    damage = gscCalculate(machamp, starmie, thunderCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(256);
  });

  test("ignoring includes Burn", () => {
    machamp.status = Statuses.BURNED;

    let damage = gscCalculate(machamp, starmie, crossChopCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(168);

    machamp.boosts[Stats.ATK] = 1;
    damage = gscCalculate(machamp, starmie, crossChopCrit, field);
    expect(damage).toHaveLength(39);
    expect(Math.max(...damage)).toEqual(127);
  });
});

test("sanity check", () => {
  const doubleEdge = new Move({ name: "Double-Edge", gen });

  const snorlax = new Pokemon({ name: "Snorlax", gen });
  const zapdos = new Pokemon({ name: "Zapdos", gen });
  let damage = gscCalculate(snorlax, zapdos, doubleEdge, field);
  expect(damage).toHaveLength(39);
  expect(Math.max(...damage)).toEqual(180);

  const umbreon = new Pokemon({ name: "Umbreon", gen });
  const blissey = new Pokemon({ name: "Blissey", gen });
  damage = gscCalculate(umbreon, blissey, doubleEdge, field);
  expect(damage).toHaveLength(39);
  expect(Math.max(...damage)).toEqual(196);
});

test("Pursuit doubles after variance", () => {
  const tyranitar = new Pokemon({ name: "Tyranitar", gen });
  const misdreavus = new Pokemon({ name: "Misdreavus", gen });
  const pursuit = new Move({ name: "Pursuit", gen });

  expect(gscCalculate(tyranitar, misdreavus, pursuit, field)).toMatchSnapshot();

  misdreavus.switchedOut = true;
  expect(gscCalculate(tyranitar, misdreavus, pursuit, field)).toMatchSnapshot();
});

test("Explosion halves Defense", () => {
  const cloyster = new Pokemon({ name: "Cloyster", gen });
  const snorlax = new Pokemon({ name: "Snorlax", gen });
  const explosion = new Move({ name: "Explosion", gen });
  const damage = gscCalculate(cloyster, snorlax, explosion, field);
  expect(damage).toHaveLength(39);
  expect(Math.max(...damage)).toEqual(542);
});

test("Hidden Power is physical for certain types", () => {
  const marowak = new Pokemon({
    name: "Marowak",
    item: "Thick Club",
    ivs: [15, 13, 13, 15, 15, 15],
    gen
  });
  const exeggutor = new Pokemon({ name: "Exeggutor", gen });
  const hiddenPower = new Move({ name: "Hidden Power", gen });
  const damage = gscCalculate(marowak, exeggutor, hiddenPower, field);
  expect(damage).toHaveLength(39);
  expect(Math.max(...damage)).toEqual(452);
});

test("Light Ball doubles Special Attack", () => {
  const pikachu = new Pokemon({
    name: "Pikachu",
    item: "Light Ball",
    gen
  });
  const zapdos = new Pokemon({ name: "Zapdos", gen });
  const thunder = new Move({ name: "Thunder", gen });
  const damage = gscCalculate(pikachu, zapdos, thunder, field);
  expect(damage).toHaveLength(39);
  expect(Math.max(...damage)).toEqual(219);
});

test("Solar Beam is reduced by rain", () => {
  const houndoom = new Pokemon({ name: "Houndoom", gen });
  const cloyster = new Pokemon({ name: "Cloyster", gen });
  const solarBeam = new Move({ name: "Solar Beam", gen });

  const damage = gscCalculate(houndoom, cloyster, solarBeam, field);
  expect(damage).toHaveLength(39);
  expect(Math.max(...damage)).toEqual(342);

  const sandDamage = gscCalculate(houndoom, cloyster, solarBeam, sand);
  expect(sandDamage).toHaveLength(39);
  expect(Math.max(...sandDamage)).toEqual(342);

  const rainDamage = gscCalculate(houndoom, cloyster, solarBeam, rain);
  expect(rainDamage).toHaveLength(39);
  expect(Math.max(...rainDamage)).toEqual(170);
});

test("Reversal and Flail have no damage variance", () => {
  const reversal = new Move({ name: "Reversal", gen });
  const flail = new Move({ name: "Flail", gen });
  const heracross = new Pokemon({
    name: "Heracross",
    currentHp: 1,
    gen
  });
  const skarmory = new Pokemon({ name: "Skarmory", gen });
  expect(gscCalculate(heracross, skarmory, reversal, field)).toEqual([235]);
  expect(gscCalculate(heracross, skarmory, flail, field)).toEqual([78]);
});

test("Rain and Sun affect Fire-type and Water-type damage", () => {
  const moltres = new Pokemon({ name: "Moltres", gen });
  const suicune = new Pokemon({ name: "Suicune", gen });
  const fireBlast = new Move({ name: "Fire Blast", gen });
  const hydroPump = new Move({ name: "Hydro Pump", gen });
  const doubleEdge = new Move({ name: "Double-Edge", gen });

  const sunBoosted = gscCalculate(moltres, suicune, fireBlast, sun);
  expect(sunBoosted).toHaveLength(39);
  expect(Math.max(...sunBoosted)).toEqual(121);

  const sunReduced = gscCalculate(moltres, suicune, fireBlast, rain);
  expect(sunReduced).toHaveLength(39);
  expect(Math.max(...sunReduced)).toEqual(40);

  const rainBoosted = gscCalculate(suicune, moltres, hydroPump, rain);
  expect(rainBoosted).toHaveLength(39);
  expect(Math.max(...rainBoosted)).toEqual(470);

  const rainReduced = gscCalculate(suicune, moltres, hydroPump, sun);
  expect(rainReduced).toHaveLength(39);
  expect(Math.max(...rainReduced)).toEqual(156);

  const rainUnboosted = gscCalculate(moltres, suicune, doubleEdge, rain);
  expect(rainUnboosted).toHaveLength(39);
  expect(Math.max(...rainUnboosted)).toEqual(92);

  const sunUnboosted = gscCalculate(moltres, suicune, doubleEdge, sun);
  expect(sunUnboosted).toHaveLength(39);
  expect(Math.max(...sunUnboosted)).toEqual(92);
});

test("Type boosting items increase damage by 10%", () => {
  const moltres = new Pokemon({
    name: "Moltres",
    item: "Charcoal",
    gen
  });
  const snorlax = new Pokemon({ name: "Snorlax", gen });
  const fireBlast = new Move({ name: "Fire Blast", gen });
  const damage = gscCalculate(moltres, snorlax, fireBlast, field);
  expect(damage).toHaveLength(39);
  expect(Math.max(...damage)).toEqual(186);
});

test("Type immunity", () => {
  const snorlax = new Pokemon({ name: "Snorlax", gen });
  const gengar = new Pokemon({ name: "Gengar", gen });
  const bodySlam = new Move({ name: "Body Slam", gen });
  expect(gscCalculate(snorlax, gengar, bodySlam, field)).toEqual([0]);
});
