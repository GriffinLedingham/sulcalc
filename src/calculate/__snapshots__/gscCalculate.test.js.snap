// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gscCalculate() Pursuit doubles after variance 1`] = `
Array [
  97,
  97,
  97,
  98,
  98,
  99,
  99,
  100,
  100,
  101,
  101,
  101,
  102,
  102,
  103,
  103,
  104,
  104,
  105,
  105,
  105,
  106,
  106,
  107,
  107,
  108,
  108,
  109,
  109,
  109,
  110,
  110,
  111,
  111,
  112,
  112,
  113,
  113,
  114,
]
`;

exports[`gscCalculate() Pursuit doubles after variance 2`] = `
Array [
  194,
  194,
  194,
  196,
  196,
  198,
  198,
  200,
  200,
  202,
  202,
  202,
  204,
  204,
  206,
  206,
  208,
  208,
  210,
  210,
  210,
  212,
  212,
  214,
  214,
  216,
  216,
  218,
  218,
  218,
  220,
  220,
  222,
  222,
  224,
  224,
  226,
  226,
  228,
]
`;
